internal/exec/stages/disks/luks.go

```diff
@@ -0,0 +1,240 @@
+type Tang struct {
+	URL        string `json:"url"`
+	Thumbprint string `json:"thp,omitempty"`
+}
+
+// https://github.com/latchset/clevis/blob/master/README.md#pin-shamir-secret-sharing
+type Pin struct {
+	Tpm  bool
+	Tang []Tang
+}
+
+func (p Pin) MarshalJSON() ([]byte, error) {
+	if p.Tpm {
+		return json.Marshal(&struct {
+			Tang []Tang   `json:"tang,omitempty"`
+			Tpm  struct{} `json:"tpm2"`
+		}{
+			Tang: p.Tang,
+			Tpm:  struct{}{},
+		})
+	} else {
+		return json.Marshal(&struct {
+			Tang []Tang `json:"tang"`
+		}{
+			Tang: p.Tang,
+		})
+	}
+}
+
+type Clevis struct {
+	Pins      Pin `json:"pins"`
+	Threshold int `json:"t"`
+}
+
+// Initially tested generating keyfiles via dd'ing to a file from /dev/urandom
+// however while cryptsetup had no problem with these keyfiles clevis seemed to
+// die on them while keyfiles generated via openssl rand -hex would work...
+func randHex(length int) (string, error) {
+	bytes := make([]byte, length)
+	// On older kernels this could block indefinitely but there's nothing
+	// that we can do about it; we don't want to use earlyrand
+	if _, err := rand.Read(bytes); err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(bytes), nil
+}
+
+func (s *stage) createLuks(config types.Config) error {
+	if len(config.Storage.Luks) == 0 {
+		return nil
+	}
+	s.Logger.PushPrefix("createLuks")
+	defer s.Logger.PopPrefix()
+
+	for _, luks := range config.Storage.Luks {
+		// TODO: allow Ignition generated KeyFiles for
+		// non-clevis devices that can be persisted.
+		// TODO: create devices in parallel.
+		// track whether Ignition creates the KeyFile
+		// so that it can be removed
+		var ignitionCreatedKeyFile bool
+		// create keyfile inside of tmpfs, it will be copied to the
+		// sysroot by the files stage
+		os.MkdirAll(distro.LuksInitramfsKeyFilePath, 0700)
+		keyFilePath := filepath.Join(distro.LuksInitramfsKeyFilePath, luks.Name)
+		if !strings.Contains(keyFilePath, distro.LuksInitramfsKeyFilePath) {
+			return fmt.Errorf("path traversal detected in device name")
+		}
+		if util.NilOrEmpty(luks.KeyFile.Source) {
+			// create a keyfile
+			key, err := randHex(4096)
+			if err != nil {
+				return fmt.Errorf("generating keyfile: %v", err)
+			}
+			if err := ioutil.WriteFile(keyFilePath, []byte(key), 0400); err != nil {
+				return fmt.Errorf("creating keyfile: %v", err)
+			}
+			ignitionCreatedKeyFile = true
+		} else {
+			mode := 0400
+			f := types.File{
+				Node: types.Node{
+					Path: keyFilePath,
+				},
+				FileEmbedded1: types.FileEmbedded1{
+					Contents: luks.KeyFile,
+					Mode:     &mode,
+				},
+			}
+			fetchOps, err := s.Util.PrepareFetches(s.Util.Logger, f)
+			if err != nil {
+				return fmt.Errorf("failed to resolve keyfile %q: %v", f.Path, err)
+			}
+			for _, op := range fetchOps {
+				msg := "writing file %q"
+				if err := s.Util.Logger.LogOp(
+					func() error {
+						return s.Util.PerformFetch(op)
```

Comment:
This doesn't set permissions.

Reply:
These keyfiles are all temporary and will be re-written inside of the files stage (and not having permissions set doesn't break anything). I'll drop the mode from the `types.File` declaration.